drivers/espi/espi_mchp_xec.c

```diff
@@ -656,6 +661,51 @@ static int espi_xec_flash_write(const struct device *dev,
 	return 0;
 }
 
+static int espi_xec_flash_erase(const struct device *dev,
+				struct espi_flash_packet *pckt)
+{
+	int ret;
+	uint32_t err_mask = MCHP_ESPI_FC_STS_IBERR |
+			MCHP_ESPI_FC_STS_OVRUN |
+			MCHP_ESPI_FC_STS_FAIL |
+			MCHP_ESPI_FC_STS_BADREQ;
+
+	struct espi_xec_data *data = (struct espi_xec_data *)(dev->driver_data);
+
+	LOG_DBG("%s", __func__);
+
+	if (!(ESPI_FC_REGS->STS & MCHP_ESPI_FC_STS_CHAN_EN)) {
+		LOG_ERR("Flash channel is disabled");
+		return -EIO;
+	}
+
+	if ((ESPI_FC_REGS->CFG & MCHP_ESPI_FC_CFG_BUSY)) {
+		LOG_ERR("Flash channel is busy");
+		return -EBUSY;
+	}
+
+	ESPI_FC_REGS->FL_ADDR_MSW = 0;
+	ESPI_FC_REGS->FL_ADDR_LSW = pckt->flash_addr;
+	ESPI_FC_REGS->XFR_LEN = ESPI_FLASH_ERASE_DUMMY;
```

Comment:
Before starting I always clear status in the block.  This code knows the controller is enabled and not busy so it's good practice to clear the status in the block and in the next level the GIRQ FC bit.

Reply:
Is something like "ESPI_FC_REGS->STS = MCHP_ESPI_FC_STS_DONE" enough? Or do you advise me to read the status register and write the same value (so that all set bits are cleared)? By GIRQ FC bit, do you mean clearing the interrupt in MCHP_GIRQ_SRC using MCHP_ESPI_FC_GIRQ_POS?

Reply:
Looks good and makes sense. In fact, i had seen an issue where subsequent writes were not going through if a previous write had failed. This should fix it. I guess this change needs to be made to write and read functions as well.

Reply:
Yes, I would add the status clear sequence before enabling the interrupt in all the functions using interrupts.

Reply:
Now when i look deeper, I am wondering why FC needs to be cleared in the MCHP_GIRQ_SRC register. The ISR would anyway fire if the FC bit gets set, correct? And the bit gets cleared in the ISR. By manually clearing SRC, we are trying to cover a case where the ISR doesn't trigger even if the bit is set... not sure if we will have such a situation if hardware behaves as expected.

Reply:
FC has status and interrupt enable registers. The interrupt enable register only allows two event to cause an interrupt: FC done and FC channel enable change. All the other FC status bits cause DONE to be set. DONE and FC Enable Change status are bit-wise AND'd with their respective bits in the FC interrupt enabled then OR'd together to produce one interrupt signal going to GIRQ19 FC bit. What HW events clear these bits?  ESPI_RESET# assertion will clear FC interrupt enable and status bits but not GIRQ19. GIRQ status bits are only cleared on chip reset. My recommendations is: Create a small static function that clears FC status and GIR19 FC status in that order. Add a call to this new function just before the call to espi_init_flash in the DEVICE_AND_API_INIT code. The driver will then clear FC status before enabling FC interrupt in espi_init_flash. This change will make sure the driver starts with FC status properly cleared in the FC registers and GIRQ19 and should not require modification of the FC read/write routines.  There is a function named espi_init_flash looks tempting to modify but it is called in two other places. It's called in the ESPI_RESET# ISR because ESPI_RESET# assertion clears FC registers and you want the FC interrupt enabled quickly after ESPI_RESET# de-asserts. HW generates an interrupt on both edges of ESPI_RESET#. The FC ISR also calls espi_init_flash for the case of the eSPI master setting FC Enable making sure the FC DONE interrupt is enabled for later use.

Reply:
Not sure i understand. There are 2 things here.
1) Clearing ESPI_FC_REGS->STS - This needs to be done inside read/write/erase since we need to account for a previous failure in read/write/erase. Doing this just once in some init function will not be sufficient.
2) Clearing FC in MCHP_GIRQ_SRC - The driver is currently clearing the FC bit in GIRQ19 after servicing the ISR. We may need to clear it once in some init function to begin with a clean slate but i don't think we need to do it every time in read/write/erase since the ISR would have anyway done it.

Reply:
@scottwcpg Please respond

Reply:
All status including the peripheral and GIRQ bits should be cleared in the driver initialization routine.  This makes sure the peripheral and GIRQ bits are synchronized as both sets of bits are R/W1C.  Clearing the peripheral bits does not clear the GIRQ bits (GIRQ bits are latched).
You can choose to only clear the FC status in the read/write/erase routines as long as the GIRQ has been handled by the ISR but this assume driver initialization has cleared both.